### PR TITLE
Faircore4eosc

### PR DESCRIPTION
**Pull request by alejandromumo**
Note to reviewers: 4.2k changes come from the vocabulary for programming languages.

closes https://github.com/inveniosoftware/invenio-swh/issues/20

### CODE REVIEW

File: invenio_rdm_records/resources/serializers/bibtex/schema.py
Comment (alejandromumo) [inline invenio_rdm_records/resources/serializers/bibtex/schema.py:1]: I think this change makes sense to be moved into a separate PR, it is something that I discussed briefly with @lnielsen before migration.

File: invenio_rdm_records/contrib/codemeta/signals.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+"""Software Heritage signals."""
+
+from blinker import Namespace
+
+_signals = Namespace()
+
+post_publish_signal = _signals.signal("swh-record-post-publish")
```
Comment (slint) [inline invenio_rdm_records/contrib/codemeta/signals.py:13]:
```suggestion
post_publish_signal = _signals.signal("record-post-publish")
```

File: invenio_rdm_records/contrib/codemeta/signals.py
Comment (slint) [inline invenio_rdm_records/contrib/codemeta/signals.py:1]: I would also move this at the top-level or inside `invenio_rdm_records/services/signals.py`

File: invenio_rdm_records/resources/serializers/__init__.py [hunk 2]
```diff
@@ -42,4 +43,6 @@
     "StringCitationSerializer",
     "UIJSONSerializer",
     "DCATSerializer",
+    "CFFSerializer",
```
Comment (slint) [inline invenio_rdm_records/resources/serializers/__init__.py:46]: minor: I think this is not imported anywhere above.

File: invenio_rdm_records/resources/serializers/codemeta/__init__.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            list_schema_cls=BaseListSchema,
+            schema_kwargs={"dumpers": [CodemetaDumper()]},  # Order matters
+            **options
+        )
+
+
+# https://raw.githubusercontent.com/codemeta/codemeta/2.0-rc/codemeta.jsonld
+# https://codemeta.github.io/codemeta-generator/
+# https://github.com/ropensci/codemetar/blob/main/codemeta.json
+# https://github.com/digitalbazaar/pyld
```
Comment (slint) [inline invenio_rdm_records/resources/serializers/codemeta/__init__.py:34]: minor: leftovers? Or if some are relevant maybe add in the docstring of the schema/serializer classes

File: invenio_rdm_records/resources/serializers/codemeta/schema.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    def get_type(self, obj):
+        """Get type.
+
+        For CodeMeta, only "SoftwareSourceCode" types are allowed.
+        """
+        read_type = super().get_type(obj)
+        if "SoftwareSourceCode" not in read_type:
+            raise ValidationError(
+                "Record is not CodeMeta serializable. It must be of type software."
+            )
```
Comment (slint) [inline invenio_rdm_records/resources/serializers/codemeta/schema.py:38]: minor: since we're not using `CodeMeta` as input, I wouldn't raise a validation error here... I would still actually serialize with the original type, and if there's some limitation about the type, enforce it higher up the call chain.

File: invenio_rdm_records/resources/serializers/codemeta/schema.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    )
+    funding = fields.Method("get_funding")
+    embargoDate = fields.Method("get_embargo_date", data_key="embargoDate")
+
+    def get_type(self, obj):
+        """Get type.
+
+        For CodeMeta, only "SoftwareSourceCode" types are allowed.
+        """
+        read_type = super().get_type(obj)
```
Comment (slint) [inline invenio_rdm_records/resources/serializers/codemeta/schema.py:34]:
```suggestion
        _type = super().get_type(obj)
```

nit: wasn't sure what the `read_` part meant :)

File: invenio_rdm_records/resources/serializers/codemeta/schema.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        if "SoftwareSourceCode" not in read_type:
+            raise ValidationError(
+                "Record is not CodeMeta serializable. It must be of type software."
+            )
+        return read_type
+
+    def get_funding(self, obj):
```
Comment (slint) [inline invenio_rdm_records/resources/serializers/codemeta/schema.py:41]: minor/shelve: we could move this up in the parent schema

File: invenio_rdm_records/resources/serializers/schemaorg/schema.py [hunk 7]
```diff
@@ -187,7 +187,7 @@ class SchemaorgSchema(BaseSerializerSchema, CommonFieldsMixin):
     dateModified = fields.Method("get_modification_date")
     temporal = fields.Method("get_dates")
     inLanguage = fields.Method("get_language")
-    contentSize = fields.Method("get_size")
```
Comment (slint) [inline invenio_rdm_records/resources/serializers/schemaorg/schema.py:190]: minor: This is used in the `MediaObject` type: https://schema.org/MediaObject

we can keep it as a duplicate (similar to `creator/author`)

**Review by slint — COMMENTED** [review]
WIP review

File: invenio_rdm_records/resources/serializers/codemeta/__init__.py (hunk 3 above)
Comment (alejandromumo) [reply invenio_rdm_records/resources/serializers/codemeta/__init__.py:34]: Definitely leftovers 😓

File: invenio_rdm_records/resources/serializers/__init__.py (hunk 2 above)
Comment (alejandromumo) [reply invenio_rdm_records/resources/serializers/__init__.py:46]: Leftover from the CFF serializer implementation, thanks!

File: invenio_rdm_records/resources/serializers/codemeta/schema.py (hunk 6 above)
Comment (alejandromumo) [reply invenio_rdm_records/resources/serializers/codemeta/schema.py:41]: I added it here since codemeta schema re-defines funding: 

https://raw.githubusercontent.com/codemeta/codemeta/2.0/codemeta.jsonld

Now that you mention it, their [docs](https://codemeta.github.io/terms/) say the type of `funding` is  "Text"  and describe it as " Funding source (e.g. specific grant)". So we should return a string here, wdyt?

Comment (tmorrell) [reply invenio_rdm_records/resources/serializers/codemeta/schema.py:29]: This has been an incredibly long-running issue....hopefully it will be fixed at some point https://github.com/codemeta/codemeta/pull/274
